impl/cluster/src/main/java/org/pentaho/big/data/impl/cluster/NamedClusterImpl.java

```diff
@@ -598,37 +608,85 @@ public String toXmlForEmbed( String rootTag ) {
-      LOGGER.error( "Could not parse embedded cluster xml" + e1.toString() );
+      LOGGER.error( "Could not parse embedded cluster xml", e1 );
       return "";
     }
   }
 
+  private Element createSiteFileChildren( Document doc, ArrayList<NamedClusterSiteFile> siteFiles ) {
+    Element children = doc.createElement( CHILDREN );
+    int index = 0;
+    for ( NamedClusterSiteFile sitefile : siteFiles ) {
+      Element siteChildren = doc.createElement( CHILDREN );
+      siteChildren
+        .appendChild( createChildElement( doc, "siteFileContents", UPPER_STRING, sitefile.getSiteFileContents(), null ) );
+      siteChildren
+        .appendChild( createChildElement( doc, "siteFileName",  UPPER_STRING, sitefile.getSiteFileName(), null ) );
+      children.appendChild( createChildElement( doc, String.valueOf( index++ ),  UPPER_STRING, "", siteChildren ) );
+    }
+    return children;
+  }
+
   public NamedCluster fromXmlForEmbed( Node node ) {
     NamedClusterImpl returnCluster = this.clone();
-    List<Node> fields = XMLHandler.getNodes( node, "child" );
+    List<Node> fields = XMLHandler.getNodes( node, CHILD );
     for ( Node field: fields ) {
-      String fieldName = XMLHandler.getTagValue( field, "id" );
-      String fieldValue = XMLHandler.getTagValue(  field, "value" );
-      if ( fieldName.toLowerCase().contains( "password" ) ) {
-        fieldValue = decodePassword( fieldValue );
+      String fieldName = XMLHandler.getTagValue( field, ID );
+      Object fieldValue = null;
+
+      if ( "siteFiles".equals( fieldName ) ) {
+        fieldValue = unmarshallSiteFileNode( field );
+      } else {
+        String stringValue = XMLHandler.getTagValue( field, VALUE );
+        if ( fieldName.toLowerCase().contains( "password" ) ) {
+          stringValue = decodePassword( stringValue );
+        }
+        fieldValue = stringValue;
       }
       try {
         BeanUtils.setProperty( returnCluster, fieldName, fieldValue );
       } catch ( IllegalAccessException | InvocationTargetException e ) {
-        LOGGER.error( "Could not serialize NamedCluster to xml: " + e.toString() );
+        LOGGER.error( "Could not set field " + fieldName + " in NamedCluster", e );
       }
     }
     return returnCluster;
   }
 
-  private Node createChildElement( Document doc, String elementName, String elementType, String elementValue ) {
-    Element childNode = doc.createElement( "child" );
-    childNode.appendChild( createTextNode( doc, "id", elementName ) );
-    childNode.appendChild( createTextNode( doc, "value", elementValue ) );
+  private Object unmarshallSiteFileNode( Node field ) {
+    ArrayList<NamedClusterSiteFile> namedClusterSiteFiles = new ArrayList<>();
+    Node siteFileWrapper = XMLHandler.getSubNode( field, CHILDREN );
+    if ( siteFileWrapper != null ) {
+      List<Node> siteFilesNode = XMLHandler.getNodes( siteFileWrapper, CHILD );
+      for ( Node siteFile : siteFilesNode ) {
```

Comment:
Can you extract some of this to not be quite so deeply nested?  Once I hit level four of a nesting my eyes start to glaze over.